File: app/src/main/java/com/example/waveplayer/fragments/FragmentSongs.java
package com.example.waveplayer.fragments;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavDirections;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.waveplayer.R;
import com.example.waveplayer.ViewModelUserPickedPlaylist;
import com.example.waveplayer.activity_main.ActivityMain;
import com.example.waveplayer.activity_main.DialogFragmentAddToPlaylist;
import com.example.waveplayer.activity_main.ViewModelActivityMain;
import com.example.waveplayer.databinding.RecyclerViewSongListBinding;
import com.example.waveplayer.random_playlist.Song;

import java.util.ArrayList;
import java.util.List;

import static com.example.waveplayer.activity_main.DialogFragmentAddToPlaylist.BUNDLE_KEY_ADD_TO_PLAYLIST_SONG;

public class FragmentSongs extends Fragment implements
        RecyclerViewAdapterSongs.ListenerCallbackSongs {

    private RecyclerViewSongListBinding binding;

    private ViewModelActivityMain viewModelActivityMain;
    private ViewModelUserPickedPlaylist viewModelUserPickedPlaylist;

    private BroadcastReceiver broadcastReceiver;

    private RecyclerView recyclerViewSongs;
    private RecyclerViewAdapterSongs recyclerViewAdapterSongs;

    private SearchView.OnQueryTextListener onQueryTextListenerSearch;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        viewModelUserPickedPlaylist =
                new ViewModelProvider(requireActivity()).get(ViewModelUserPickedPlaylist.class);
        viewModelActivityMain =
                new ViewModelProvider(requireActivity()).get(ViewModelActivityMain.class);
        binding = RecyclerViewSongListBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ActivityMain activityMain = (ActivityMain) requireActivity();
        activityMain.hideKeyboard(view);
        viewModelActivityMain.setActionBarTitle(getResources().getString(R.string.songs));
        viewModelActivityMain.showFab(false);
        viewModelUserPickedPlaylist.setUserPickedPlaylist(activityMain.getMasterPlaylist());
        setUpBroadcastReceiver();
    }

    private void setUpBroadcastReceiver() {
        final ActivityMain activityMain = (ActivityMain) requireActivity();
        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addCategory(Intent.CATEGORY_DEFAULT);
        intentFilter.addAction(activityMain.getResources().getString(
                R.string.broadcast_receiver_action_service_connected));
        intentFilter.addAction(activityMain.getResources().getString(
                R.string.broadcast_receiver_action_on_create_options_menu));
        broadcastReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                String action = intent.getAction();
                if (action != null) {
                    if (action.equals(getResources().getString(
                            R.string.broadcast_receiver_action_on_create_options_menu))) {
                        setUpToolbar();
                    } else if (action.equals(getResources().getString(
                            R.string.broadcast_receiver_action_service_connected))) {
                        setUpRecyclerView();
                    }
                }
            }
        };
        activityMain.registerReceiver(broadcastReceiver, intentFilter);
    }

    private void setUpToolbar() {
        ActivityMain activityMain = (ActivityMain) requireActivity();
        Toolbar toolbar = activityMain.findViewById(R.id.toolbar);
        Menu menu = toolbar.getMenu();
        if (menu != null) {
            menu.getItem(ActivityMain.MENU_ACTION_RESET_PROBS_INDEX).setVisible(true);
            menu.getItem(ActivityMain.MENU_ACTION_SEARCH_INDEX).setVisible(true);
            MenuItem itemSearch = menu.findItem(R.id.action_search);
            if (itemSearch != null) {
                onQueryTextListenerSearch = new SearchView.OnQueryTextListener(){

                    @Override
                    public boolean onQueryTextSubmit(String query) {
                        return false;
                    }

                    @Override
                    public boolean onQueryTextChange(String newText) {
                        RecyclerView recyclerViewSongs = activityMain.findViewById(R.id.recycler_view_song_list);
                        if (recyclerViewSongs != null) {
                            RecyclerViewAdapterSongs recyclerViewAdapterSongs =
                                    (RecyclerViewAdapterSongs) recyclerViewSongs.getAdapter();
                            List<Song> songs = activityMain.getAllSongs();
                            List<Song> sifted = new ArrayList<>();
                            if (!newText.equals("")) {
                                for (Song song : songs) {
                                    if (song.title.toLowerCase().contains(newText.toLowerCase())) {
                                        sifted.add(song);
                                    }
                                }
                                recyclerViewAdapterSongs.updateList(sifted);
                            } else {
                                recyclerViewAdapterSongs.updateList(songs);
                            }
                        }
                        return true;
                    }
                };
                SearchView searchView = (SearchView) itemSearch.getActionView();
                searchView.setOnQueryTextListener(onQueryTextListenerSearch);
            }
        }
    }

    private void setUpRecyclerView() {
        ActivityMain activityMain = (ActivityMain) requireActivity();
        recyclerViewSongs = binding.recyclerViewSongList;
        List<Song> songs = activityMain.getAllSongs();
        if (songs != null) {
            recyclerViewAdapterSongs = new RecyclerViewAdapterSongs(
                    this, new ArrayList<>(songs));
            recyclerViewSongs.setLayoutManager(new LinearLayoutManager(recyclerViewSongs.getContext()));
            recyclerViewSongs.setAdapter(recyclerViewAdapterSongs);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        setUpToolbar();
        setUpRecyclerView();
    }

    @Override
    public boolean onMenuItemClickAddToPlaylist(Song song) {
        Bundle bundle = new Bundle();
        bundle.putSerializable(BUNDLE_KEY_ADD_TO_PLAYLIST_SONG, song);
        DialogFragmentAddToPlaylist dialogFragmentAddToPlaylist =
                new DialogFragmentAddToPlaylist();
        dialogFragmentAddToPlaylist.setArguments(bundle);
        dialogFragmentAddToPlaylist.show(getParentFragmentManager(), getTag());
        return true;
    }

    @Override
    public boolean onMenuItemClickAddToQueue(Song song) {
        ActivityMain activityMain = (ActivityMain) requireActivity();
        if (activityMain.songInProgress()) {
            activityMain.addToQueue(song.id);
        } else {
            activityMain.showSongPane();
            activityMain.addToQueue(song.id);
            if (!activityMain.isSongInProgress()) {
                activityMain.playNext();
            }
        }
        return true;
    }

    @Override
    public void onClickViewHolder(Song song) {
        ActivityMain activityMain = ((ActivityMain) requireActivity());
        if (song.equals(activityMain.getCurrentSong())) {
            activityMain.seekTo(0);
        }
        activityMain.setCurrentPlaylistToMaster();
        activityMain.clearSongQueue();
        activityMain.addToQueue(song.id);
        activityMain.playNext();
        NavDirections action = FragmentSongsDirections.actionFragmentSongsToFragmentSong();
        if (action != null) {
            NavHostFragment.findNavController(this).navigate(action);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        final ActivityMain activityMain = ((ActivityMain) requireActivity());
        View view = requireView();
        activityMain.hideKeyboard(view);
        Toolbar toolbar = activityMain.findViewById(R.id.toolbar);
        Menu menu = toolbar.getMenu();
        if (menu != null) {
            MenuItem itemSearch = menu.findItem(R.id.action_search);
            SearchView searchView = (SearchView) itemSearch.getActionView();
            searchView.setOnQueryTextListener(null);
            searchView.onActionViewCollapsed();
        }
        onQueryTextListenerSearch = null;
        recyclerViewSongs.setAdapter(null);
        recyclerViewAdapterSongs = null;
        recyclerViewSongs = null;
        activityMain.unregisterReceiver(broadcastReceiver);
        broadcastReceiver = null;
        viewModelUserPickedPlaylist = null;
        viewModelActivityMain = null;
        binding = null;
    }

}

File: app/src/main/java/com/example/waveplayer/fragments/FragmentSongs.kt
package com.example.waveplayer.fragments

import android.content.Context
import android.view.View
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import com.example.waveplayer.databinding.RecyclerViewSongListBinding
import java.util.*

class FragmentSongs : Fragment(), ListenerCallbackSongs {
    private var binding: RecyclerViewSongListBinding? = null
    private var viewModelActivityMain: ViewModelActivityMain? = null
    private var viewModelUserPickedPlaylist: ViewModelUserPickedPlaylist? = null
    private var broadcastReceiver: BroadcastReceiver? = null
    private var recyclerViewSongs: RecyclerView? = null
    private var recyclerViewAdapterSongs: RecyclerViewAdapterSongs? = null
    private var onQueryTextListenerSearch: SearchView.OnQueryTextListener? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        viewModelUserPickedPlaylist = ViewModelProvider(requireActivity()).get<ViewModelUserPickedPlaylist?>(ViewModelUserPickedPlaylist::class.java)
        viewModelActivityMain = ViewModelProvider(requireActivity()).get<ViewModelActivityMain?>(ViewModelActivityMain::class.java)
        binding = RecyclerViewSongListBinding.inflate(inflater, container, false)
        return binding.getRoot()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val activityMain: ActivityMain = requireActivity() as ActivityMain
        activityMain.hideKeyboard(view)
        viewModelActivityMain.setActionBarTitle(resources.getString(R.string.songs))
        viewModelActivityMain.showFab(false)
        viewModelUserPickedPlaylist.setUserPickedPlaylist(activityMain.getMasterPlaylist())
        setUpBroadcastReceiver()
    }

    private fun setUpBroadcastReceiver() {
        val activityMain: ActivityMain = requireActivity() as ActivityMain
        val intentFilter = IntentFilter()
        intentFilter.addCategory(Intent.CATEGORY_DEFAULT)
        intentFilter.addAction(activityMain.getResources().getString(
                R.string.broadcast_receiver_action_service_connected))
        intentFilter.addAction(activityMain.getResources().getString(
                R.string.broadcast_receiver_action_on_create_options_menu))
        broadcastReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context?, intent: Intent?) {
                val action: String = intent.getAction()
                if (action != null) {
                    if (action == resources.getString(
                                    R.string.broadcast_receiver_action_on_create_options_menu)) {
                        setUpToolbar()
                    } else if (action == resources.getString(
                                    R.string.broadcast_receiver_action_service_connected)) {
                        setUpRecyclerView()
                    }
                }
            }
        }
        activityMain.registerReceiver(broadcastReceiver, intentFilter)
    }

    private fun setUpToolbar() {
        val activityMain: ActivityMain = requireActivity() as ActivityMain
        val toolbar: Toolbar = activityMain.findViewById<Toolbar?>(R.id.toolbar)
        val menu = toolbar.menu
        if (menu != null) {
            menu.getItem(ActivityMain.Companion.MENU_ACTION_RESET_PROBS_INDEX).isVisible = true
            menu.getItem(ActivityMain.Companion.MENU_ACTION_SEARCH_INDEX).isVisible = true
            val itemSearch = menu.findItem(R.id.action_search)
            if (itemSearch != null) {
                onQueryTextListenerSearch = object : SearchView.OnQueryTextListener {
                    override fun onQueryTextSubmit(query: String?): Boolean {
                        return false
                    }

                    override fun onQueryTextChange(newText: String?): Boolean {
                        val recyclerViewSongs: RecyclerView = activityMain.findViewById<RecyclerView?>(R.id.recycler_view_song_list)
                        if (recyclerViewSongs != null) {
                            val recyclerViewAdapterSongs = recyclerViewSongs.getAdapter() as RecyclerViewAdapterSongs
                            val songs: MutableList<Song?> = activityMain.getAllSongs()
                            val sifted: MutableList<Song?> = ArrayList<Song?>()
                            if (newText != "") {
                                for (song in songs) {
                                    if (song.title.toLowerCase().contains(newText.toLowerCase())) {
                                        sifted.add(song)
                                    }
                                }
                                recyclerViewAdapterSongs.updateList(sifted)
                            } else {
                                recyclerViewAdapterSongs.updateList(songs)
                            }
                        }
                        return true
                    }
                }
                val searchView = itemSearch.actionView as SearchView
                searchView.setOnQueryTextListener(onQueryTextListenerSearch)
            }
        }
    }

    private fun setUpRecyclerView() {
        val activityMain: ActivityMain = requireActivity() as ActivityMain
        recyclerViewSongs = binding.recyclerViewSongList
        val songs: MutableList<Song?> = activityMain.getAllSongs()
        if (songs != null) {
            recyclerViewAdapterSongs = RecyclerViewAdapterSongs(
                    this, ArrayList<Song?>(songs))
            recyclerViewSongs.setLayoutManager(LinearLayoutManager(recyclerViewSongs.getContext()))
            recyclerViewSongs.setAdapter(recyclerViewAdapterSongs)
        }
    }

    override fun onResume() {
        super.onResume()
        setUpToolbar()
        setUpRecyclerView()
    }

    override fun onMenuItemClickAddToPlaylist(song: Song?): Boolean {
        val bundle = Bundle()
        bundle.putSerializable(DialogFragmentAddToPlaylist.Companion.BUNDLE_KEY_ADD_TO_PLAYLIST_SONG, song)
        val dialogFragmentAddToPlaylist = DialogFragmentAddToPlaylist()
        dialogFragmentAddToPlaylist.setArguments(bundle)
        dialogFragmentAddToPlaylist.show(parentFragmentManager, tag)
        return true
    }

    override fun onMenuItemClickAddToQueue(song: Song?): Boolean {
        val activityMain: ActivityMain = requireActivity() as ActivityMain
        if (activityMain.songInProgress()) {
            activityMain.addToQueue(song.id)
        } else {
            activityMain.showSongPane()
            activityMain.addToQueue(song.id)
            if (!activityMain.isSongInProgress()) {
                activityMain.playNext()
            }
        }
        return true
    }

    override fun onClickViewHolder(song: Song?) {
        val activityMain: ActivityMain = requireActivity() as ActivityMain
        if (song == activityMain.getCurrentSong()) {
            activityMain.seekTo(0)
        }
        activityMain.setCurrentPlaylistToMaster()
        activityMain.clearSongQueue()
        activityMain.addToQueue(song.id)
        activityMain.playNext()
        val action: NavDirections = FragmentSongsDirections.actionFragmentSongsToFragmentSong()
        val navController: NavController = NavHostFragment.findNavController(this)
        if (navController.getCurrentDestination().getId() == R.id.fragmentSongs) {
            navController.navigate(action)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        val activityMain: ActivityMain = requireActivity() as ActivityMain
        activityMain.unregisterReceiver(broadcastReceiver)
        broadcastReceiver = null
        val view = requireView()
        activityMain.hideKeyboard(view)
        val toolbar: Toolbar = activityMain.findViewById<Toolbar?>(R.id.toolbar)
        val menu = toolbar.menu
        if (menu != null) {
            val itemSearch = menu.findItem(R.id.action_search)
            val searchView = itemSearch.actionView as SearchView
            searchView.setOnQueryTextListener(null)
            searchView.onActionViewCollapsed()
        }
        onQueryTextListenerSearch = null
        recyclerViewSongs.setAdapter(null)
        recyclerViewAdapterSongs = null
        recyclerViewSongs = null
        viewModelUserPickedPlaylist = null
        viewModelActivityMain = null
        binding = null
    }
}